src/runtime/schema.ts

```diff
@@ -57,10 +57,66 @@ export class Schema {
       }
     }
     if (options.description) {
+      // The descriptions should be passed ready for assignment into this.description.
+      // TODO(cypher1): Refactor the schema construction code to do this rearrangement at the call site.
       options.description.description.forEach(desc => this.description[desc.name] = desc.pattern || desc.patterns[0]);
     }
   }
 
+  private forEachRefinement<T, V>(func: (val: V) => T) {
+    const types = [this, ...Object.values(this.fields)];
+    types.forEach(type => {
+      if (type.refinement) {
+        func(type.refinement);
+      }
+    });
+  }
+
+  getFieldParams(): Map<string, Primitive> {
+    const params = new Map<string, Primitive>();
+    this.forEachRefinement(
+      (ref: Refinement) => mergeMapInto(params, ref.getFieldParams())
+    );
+    return params;
+  }
+
+  getQueryParams(): Map<string, Primitive> {
+    const params = new Map<string, Primitive>();
+    this.forEachRefinement(
+      (ref: Refinement) => mergeMapInto(params, ref.getQueryParams())
+    );
+    return params;
+  }
+
+  toKTExpression(codeGenerator): string[] {
+    const filterExpressions = [];
+    this.forEachRefinement(
+      (ref: Refinement) => filterExpressions.push(ref.toKTExpression(codeGenerator))
+    );
+    return filterExpressions;
+  }
+
+  getQueryType(): string {
+    return this.getQueryParams().get('?');
+  }
+
+  extractRefinementFromQuery(): Schema {
+    const fields = [];
+    for (const [name, fieldType] of Object.entries(this.fields)) {
+      const field = {...fieldType};
+      field.refinement = fieldType.refinement && fieldType.refinement.extractRefinementFromQuery();
```

Comment:
Trivial: couldn't you just use `field` instead of `fieldType` on the right side? Saves a few chars... :)

Reply:
My preference is
```
const field = {...fieldType,
  refinement: fieldType.refinement && fieldType.refinement.extractRefinementFromQuery()
};
```
But I was worried that this was somewhat hard to read (and relies on js keeping the right most value for repeated keys).

I'll update to yours unless I hear back.

Reply:
> But I was worried that this was somewhat hard to read 

Possibly, and this one's probably more subjective than most readability points.
 
> and relies on js keeping the right most value for repeated keys

Not sure exactly what you mean, but it's very much guaranteed that the rhs expression will be evaluated before any assignment to the lhs, so that's safe.

Anyway, it's still a very trivial thing and totally up to you; I'm not really strongly for or agin' it.

Reply:
In some languages, the evaluation order doesn't give you that override behavior (e.g. rust has the opposite, things are set once and not overridden). 

I had to check when writing it because I wasn't sure that it was correct and wouldn't want readers to have to do the same. Thanks